### ISSUE TITLE
syscall: treat ENFILE as a temporary error, like EMFILE

### ISSUE THREAD

**Issue by lmb** [golang/go#35131]
### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

Linux

### What did you do?

One of our servers ran out of file handles (/proc/sys/fs/file-max). This led net.Listener.Accept() to return ENFILE (not EMFILE). Since ENFILE is not considered a temporary error, http.Server.Serve exited it's accept loop and returned ENFILE.

### What did you expect to see?

ENFILE should be treated as temporary if returned by accept (and possible others). http.Server should not abort its accept loop when encountering ENFILE.

### What did you see instead?

The application which received ENFILE logged the error, and continued running without a working listener / http server. If the application had followed the common `log.Fatalln(http.Serve(...))` it would have crashed.

**Comment by networkimprov** [comment]
On Windows, there's `WSAEMFILE` but according to the page below `WSAENOBUFS` is more likely, so maybe that should be added to net/error_windows.go:isConnError()

https://emptyhammock.com/blog/wsaemfile.html

**Comment by ianlancetaylor** [comment]
If we're going to change this we should change it in the syscall package, in `Errno.Temporary`, not in the net package.

**Comment by ianlancetaylor** [comment]
Looks like the check for `EMFILE` was added in https://golang.org/cl/4550112, but we somehow failed to consider checking for `ENFILE` at that time.  CC @bradfitz who wrote 4550112 in case he remembers anything.

**Comment by networkimprov** [comment]
Also `WSAEMFILE` & `WSAENOBUFS` for syscall_windows.go

cc @alexbrainman @mattn @zx2c4

**Comment by lmb** [comment]
I've updated the CL to change `Errno.Temporary` as suggested by @ianlancetaylor. Can someone else tackle the Windows part, or split it into a separate issue? I don't have access to a Windows machine for testing.

**Comment by networkimprov** [comment]
@alexbrainman @ianlancetaylor the CL for Windows has now eliminated syscall.ENFILE instead of adding WSAEMFILE & WSAENOBUFS to the .Temporary() set. I think we want the latter, and not the former.

Tuning the list of Errno constants for Windows is #32309; a CL for it should consider all the "fake" Errno constants defined in pkg syscall, and decide whether to drop them or set them to a common value.

cc @iWdGo

**Comment by ianlancetaylor** [comment]
We shouldn't touch any of the fake `Errno` constants defined for Windows in the syscall package.  The gain from that would not be worth the break in backward compatibility.

**Comment by networkimprov** [comment]
@ianlancetaylor can you endorse adding the WSA* errors I listed to .Temporary()? I believe that was the point of reopening this, but Constantin doesn't believe me :-)

https://emptyhammock.com/blog/wsaemfile.html

**Comment by ianlancetaylor** [comment]
Yes, I think that is the appropriate change here.

**Comment by networkimprov** [comment]
Ian, there is a misunderstanding in the CL thread. Can you reply there, and drop your -2?

https://golang.org/cl/208537

**Comment by networkimprov** [comment]
@alexbrainman can you explain why `WSAEMFILE` & `WSAENOBUFS` are not temporary errors?

This reference shows the latter occurs when server is loaded:
https://emptyhammock.com/blog/wsaemfile.html

**Comment by networkimprov** [comment]
@alexbrainman sorry, the link I mentioned is posted in this thread, including immediately above. Unfortunately it doesn't give the code he used, but that should be simple to recreate if nec.

**Comment by networkimprov** [comment]
This issue is blocked by lack of WSAGetLastError. Filed #38134

**Comment by networkimprov** [comment]
@djdv this is the only open issue for temporary WSA errors. The original CL author has ceased working on it. If you'd like to take over that work and include the other WSA error you found, that would be great!

See https://golang.org/doc/contribute.html

**Comment by djdv** [comment]
@networkimprov 
Thanks for that.
I'll put this on my TODO list, but I have some other things to take care of first. If someone wants to beat me to it that'd be fine. :^)
Otherwise I'll look into this soon-ish. (I'll read this thread and linked things in full later)

Specifically, I/someone need(s) to go through the WSA error list and figure out which should be considered "temporary" / what `net` considers to be "temporary".
Initial assumption being that; as long as the connection can have a successful read in subsequent calls, that it is to be considered temporary. (leaving errors that render the socket unusable as they are now)

**Comment by networkimprov** [comment]
I think you'd want empirical evidence that an error isn't a cause to drop the connection (and some probably don't occur in practice), so I don't think you'd need to review the WSA list. Let's just add the two discussed above and the one you found.

I think there are only minor scope problems with the posted CL.
